File: backend/src/services/sftpSettingsService.ts
import { getAdminSetting, setAdminSetting } from "./adminSettingsService";

/**
 * DB-stored connection config for the Adobe Campaign SFTP deployment target
 * used by sftpService.ts (fired on every Publish/Deploy — see
 * formBuilderService.publishForm), admin-managed via Configuration >
 * Deployment. Mirrors fabrixSettingsService.ts's AdminSetting-row-per-field
 * convention, minus the encryption — nothing stored here is a secret in
 * itself, just a host/username/local-file-path/remote-path, unlike FabriX's
 * bearer tokens.
 *
 * Two independent targets ("staging"/"production") are always both stored;
 * `activeEnvironment` picks which one sftpService.ts actually deploys to.
 * `privateKeyPath` is a path on whichever machine runs the backend, not the
 * key's contents — the key file itself never leaves that machine's disk and
 * is never stored in this table or committed to the repo.
 */
export type SftpEnvironment = "staging" | "production";

export interface SftpTargetConfig {
  host: string;
  port: number;
  username: string;
  privateKeyPath: string;
  remotePath: string;
}

export interface SftpDeploymentSettings {
  activeEnvironment: SftpEnvironment;
  staging: SftpTargetConfig;
  production: SftpTargetConfig;
}

const DEFAULT_PORT = 22;
const DEFAULT_REMOTE_PATH = "/incoming/ACC/Operations/LocalHR/DWF/ToProcess";

function settingKeyPrefix(environment: SftpEnvironment): string {
  return environment === "staging" ? "sftpStaging" : "sftpProduction";
}

async function getTarget(environment: SftpEnvironment): Promise<SftpTargetConfig> {
  const prefix = settingKeyPrefix(environment);
  const [host, port, username, privateKeyPath, remotePath] = await Promise.all([
    getAdminSetting(`${prefix}Host`),
    getAdminSetting(`${prefix}Port`),
    getAdminSetting(`${prefix}Username`),
    getAdminSetting(`${prefix}PrivateKeyPath`),
    getAdminSetting(`${prefix}RemotePath`),
  ]);
  return {
    host: host ?? "",
    port: Number(port) || DEFAULT_PORT,
    username: username ?? "",
    privateKeyPath: privateKeyPath ?? "",
    remotePath: remotePath ?? DEFAULT_REMOTE_PATH,
  };
}

export async function getSftpDeploymentSettings(): Promise<SftpDeploymentSettings> {
  const [activeRaw, staging, production] = await Promise.all([
    getAdminSetting("sftpActiveEnvironment"),
    getTarget("staging"),
    getTarget("production"),
  ]);
  const activeEnvironment: SftpEnvironment = activeRaw === "production" ? "production" : "staging";
  return { activeEnvironment, staging, production };
}

export async function saveSftpTarget(environment: SftpEnvironment, input: SftpTargetConfig): Promise<void> {
  const prefix = settingKeyPrefix(environment);
  await Promise.all([
    setAdminSetting(`${prefix}Host`, input.host),
    setAdminSetting(`${prefix}Port`, String(input.port || DEFAULT_PORT)),
    setAdminSetting(`${prefix}Username`, input.username),
    setAdminSetting(`${prefix}PrivateKeyPath`, input.privateKeyPath),
    setAdminSetting(`${prefix}RemotePath`, input.remotePath || DEFAULT_REMOTE_PATH),
  ]);
}

export async function setActiveSftpEnvironment(environment: SftpEnvironment): Promise<void> {
  await setAdminSetting("sftpActiveEnvironment", environment);
}

/** The currently-active environment's config, or null if that environment's
 * required fields (host/username/privateKeyPath) aren't all filled in yet —
 * sftpService.deployGeneratedFiles treats null as "not configured." */
export async function getActiveSftpTarget(): Promise<SftpTargetConfig | null> {
  const settings = await getSftpDeploymentSettings();
  const target = settings.activeEnvironment === "production" ? settings.production : settings.staging;
  if (!target.host || !target.username || !target.privateKeyPath) return null;
  return target;
}

File: backend-py/app/services/sftp_settings_service.py
"""Port of `backend/src/services/sftpSettingsService.ts`.

DB-stored connection config for the Adobe Campaign SFTP deployment target,
admin-managed via Configuration > Deployment. Mirrors
`fabrix_settings_service`'s AdminSetting-row-per-field convention, minus the
encryption — nothing stored here is a secret in itself, just a
host/username/local-file-path/remote-path.

Two independent targets ("staging"/"production") are always both stored;
`activeEnvironment` picks which one deployment actually targets.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from sqlalchemy.orm import Session

from app.services.admin_settings_service import get_admin_setting, set_admin_setting

SftpEnvironment = Literal["staging", "production"]

DEFAULT_PORT = 22
DEFAULT_REMOTE_PATH = "/incoming/ACC/Operations/LocalHR/DWF/ToProcess"


@dataclass
class SftpTargetConfig:
    host: str
    port: int
    username: str
    privateKeyPath: str
    remotePath: str


@dataclass
class SftpDeploymentSettings:
    activeEnvironment: SftpEnvironment
    staging: SftpTargetConfig
    production: SftpTargetConfig


def _setting_key_prefix(environment: SftpEnvironment) -> str:
    return "sftpStaging" if environment == "staging" else "sftpProduction"


def _get_target(db: Session, environment: SftpEnvironment) -> SftpTargetConfig:
    prefix = _setting_key_prefix(environment)
    host = get_admin_setting(db, f"{prefix}Host")
    port = get_admin_setting(db, f"{prefix}Port")
    username = get_admin_setting(db, f"{prefix}Username")
    private_key_path = get_admin_setting(db, f"{prefix}PrivateKeyPath")
    remote_path = get_admin_setting(db, f"{prefix}RemotePath")
    try:
        port_num = int(port) if port else DEFAULT_PORT
    except ValueError:
        port_num = DEFAULT_PORT
    return SftpTargetConfig(
        host=host or "",
        port=port_num or DEFAULT_PORT,
        username=username or "",
        privateKeyPath=private_key_path or "",
        remotePath=remote_path or DEFAULT_REMOTE_PATH,
    )


def get_sftp_deployment_settings(db: Session) -> SftpDeploymentSettings:
    active_raw = get_admin_setting(db, "sftpActiveEnvironment")
    staging = _get_target(db, "staging")
    production = _get_target(db, "production")
    active_environment: SftpEnvironment = "production" if active_raw == "production" else "staging"
    return SftpDeploymentSettings(activeEnvironment=active_environment, staging=staging, production=production)


def save_sftp_target(db: Session, environment: SftpEnvironment, input: SftpTargetConfig) -> None:
    prefix = _setting_key_prefix(environment)
    set_admin_setting(db, f"{prefix}Host", input.host)
    set_admin_setting(db, f"{prefix}Port", str(input.port or DEFAULT_PORT))
    set_admin_setting(db, f"{prefix}Username", input.username)
    set_admin_setting(db, f"{prefix}PrivateKeyPath", input.privateKeyPath)
    set_admin_setting(db, f"{prefix}RemotePath", input.remotePath or DEFAULT_REMOTE_PATH)


def set_active_sftp_environment(db: Session, environment: SftpEnvironment) -> None:
    set_admin_setting(db, "sftpActiveEnvironment", environment)


def get_active_sftp_target(db: Session) -> SftpTargetConfig | None:
    """The currently-active environment's config, or `None` if that
    environment's required fields (host/username/privateKeyPath) aren't all
    filled in yet. Not called from any route in this phase — ported so the
    deployment phase can use it directly."""
    settings = get_sftp_deployment_settings(db)
    target = settings.production if settings.activeEnvironment == "production" else settings.staging
    if not target.host or not target.username or not target.privateKeyPath:
        return None
    return target
